<a href="https://colab.research.google.com/github/faresayman-ai/FlyRank-Internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

In [90]:
lane = "CTR / Engagement Opportunity Scoring"
# I'm choosing this lane because SEO teams rank plenty of pages onto page 1, but ranking
# isn't the same as earning the click — a page at position #3 can still be quietly leaking
# clicks if its snippet doesn't match intent. Scoring that gap turns a huge, noisy list of
# ranking pages into a short worklist an editor can actually act on.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

In [91]:
# decision: "which page/query pairs to prioritize for CTR fixes
# actor: "SEO analyst / content editor
# cost_of_wrong_call: "wasted editor hours (false positive) or a missed click-leak (false negative)
# why_not_a_rule: "expected CTR depends on position, intent, and device — not one threshold
# claim_level: "decision-support only, no causal claims

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [88]:
import os, sys, subprocess
import pandas as pd

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "-r", "requirements.txt"], check=True)
else:
    # find the repo root from wherever this kernel started
    while not os.path.isdir("data/raw") and os.getcwd() != "/":
        os.chdir("..")

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")
for col in ['trend_direction', 'impression_tier', 'age_tier', 'freshness_tier']:
    print(df.groupby(col)['ctr'].agg(['mean', 'count']))
    print()

                     mean  count
trend_direction                 
down             0.324138  16262
flat             1.376753   1152
new              1.296521   2236
stable           0.517321   5962
up               0.565533   4388

                     mean  count
impression_tier                 
excellent        0.312662   1078
good             0.308314   7205
low              0.937000  11248
moderate         0.212453  10469

              mean  count
age_tier                 
181-365   0.817498  11368
31-90     0.338577    492
365+      0.281184   6360
91-180    0.345821  11780

                    mean  count
freshness_tier                 
0-30            0.609021  20480
181+            3.693276    174
31-90           0.117543    175
91-180          0.238367   9171



## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

**I can say:** CTR varies meaningfully by impression_tier and age_tier (large-sample groups), low-impression-tier pages show a higher average CTR than excellent-tier pages.

**I can't say:** Any tier causes a CTR difference (no causal test run). This model predicts what an individual page will do next.

**I observed that:** In the freshness_tier, the 181+ and 31-90 groups have sample sizes below 180 rows, making their measured CTR averages statistically unstable despite their high values.

## Self-check

Before you submit, confirm each line honestly:

- [✔] Every section above is filled — markdown thinking AND the code that backs it
- [✔] The notebook runs top to bottom with no errors (Runtime → Run all)
- [✔] No client names, URLs, or private queries anywhere
- [✔] My claims use careful words: observed, measured, directional, decision-support
- [✔] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.